src/mmapbitarray.c

```diff
@@ -390,7 +390,7 @@ static inline int _assert_comparable(MBArray * array1, MBArray * array2)
         return 1;
     }
 
-    if (memcmp((char *)array1->vector, (char *)array2->vector, array1->preamblebytes)) {
+    if (memcmp((char *)array1->vector, (char *)array2->vector, MBAMAGICSIZE + sizeof(BTYPE) + sizeof(int32_t))) {
```

Comment:
This doesn't make much sense. The amount of bytes you're comparing is _19_ (9 + 8 + 2), which isn't a multiple of the element size.

What exactly are you trying to achieve here ?

Reply:
It actually is 9 + 4 + 4, but regardless, it checks that the first 9 bytes are equal (they both should be "MBITARRAY"). The next 4 bytes contains a `uint32_t` that indicates the number of bits of the bit array (`num_bits` in the Cython part). And the following 4 bytes contains a `int32_t` that indicates the number of bytes of the header.

I'm basically comparing what `_initialize_file` writes, skipping the header itself and the ending null character that I just don't care:

```c
int _initialize_file(int fd, size_t end, BTYPE num_bits, const char * header, int32_t header_len)
{
    unsigned char zero = 0;
    errno = 0;
    lseek(fd, 0, SEEK_SET);
    if (write(fd, MBAMAGIC, MBAMAGICSIZE) != MBAMAGICSIZE) {
        return 1;
    }
    if (write(fd, &num_bits, sizeof(BTYPE)) != sizeof(BTYPE)) {
        return 1;
    }
    if (write(fd, &header_len, sizeof(header_len)) != sizeof(header_len)) {
        return 1;
    }
    if (header_len) {
        if (write(fd, header, header_len) != header_len) {
            return 1;
        }
    }

    lseek(fd, end, SEEK_SET);
    if (write(fd, &zero, 1) != 1) {
        return 1;
    }

    if (errno) {
        return 1;
    }
    return 0;
}
```

Reply:
Ugh. This is absolutely sloppy. The header should be in a byte buffer instead of embedded directly in the vector. But whatever, this is not the PR for that.